lark/parsers/lalr_parser.py

```diff
@@ -35,84 +35,116 @@ def parse(self, *args):
         return self.parser.parse(*args)
 
 
-class _Parser:
-    def __init__(self, parse_table, callbacks, debug=False):
-        self.parse_table = parse_table
-        self.callbacks = callbacks
-        self.debug = debug
+class ParserState:
+    __slots__ = 'parse_table', 'lexer', 'callbacks', 'start', 'state_stack', 'value_stack', 'start_state', 'end_state', 'states'
 
-    def parse(self, seq, start, set_state=None, value_stack=None, state_stack=None):
-        token = None
-        stream = iter(seq)
-        states = self.parse_table.states
-        start_state = self.parse_table.start_states[start]
-        end_state = self.parse_table.end_states[start]
+    def __init__(self, parse_table, lexer, callbacks, start, state_stack=None, value_stack=None):
+        self.parse_table = parse_table
 
-        state_stack = state_stack or [start_state]
-        value_stack = value_stack or []
+        self.start_state = self.parse_table.start_states[start]
+        self.end_state = self.parse_table.end_states[start]
+        self.states = self.parse_table.states
 
-        if set_state: set_state(start_state)
+        self.lexer = lexer
+        self.callbacks = callbacks
+        self.start = start
+        self.state_stack = state_stack or [self.start_state]
+        self.value_stack = value_stack or []
+
+    @property
+    def position(self):
+        return self.state_stack[-1]
+
+    def __copy__(self):
+        return type(self)(
+            self.parse_table,
+            self.lexer, # XXX copy
+            self.callbacks,
+            self.start,
+            list(self.state_stack),
```

Comment:
I find `list` here a bit confusing as a replacement for copy. I assume that you want to avoid copying the states with `debug=True`, but just using `copy` is more clear.

Reply:
I think they are functionally equivalent. Why does `debug=True` matter?

Reply:
They are functionally equivalant, but I just think that `copy` is more clear. `debug=True` matters when you would use `deepcopy` instead, since that would deepcopy the state objects, which is not what we want. (We are using `deepcopy` one line down)